### PR TITLE
Feature/post crud

### PR DESCRIPTION
**Pull request by Bijelic03**

### CODE REVIEW

File: src/main/java/com/internship/socialnetwork/controller/PostController.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    public ResponseEntity<List<PostDto>> getAll() {
+        return ResponseEntity.ok(postService.getAll());
+    }
+
+    @GetMapping("/{id}")
+    public ResponseEntity<PostDto> getOne(@RequestParam Long id) {
+        return ResponseEntity.ok(postService.get(id));
+    }
+
+    @GetMapping("/user/{id}")
```
Comment (markonikolic-nqode) [inline src/main/java/com/internship/socialnetwork/controller/PostController.java]: Move this to userController, we should use plural as RESTful names.

File: src/main/java/com/internship/socialnetwork/controller/PostController.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    public ResponseEntity<List<PostDto>> getAllByAuthorId(@RequestParam Long id) {
+        return ResponseEntity.ok(postService.getAllByAuthorId(id));
+    }
+
+    @PostMapping
+    public ResponseEntity<PostDto> create(@Valid @RequestBody PostDto postDto) {
+        return ResponseEntity.ok(postService.create(postDto));
```
Comment (markonikolic-nqode) [inline src/main/java/com/internship/socialnetwork/controller/PostController.java]: Create should return 201

File: src/main/java/com/internship/socialnetwork/repository/PostRepository.java [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+import org.springframework.data.repository.query.Param;
+import org.springframework.stereotype.Repository;
+
+import java.util.List;
+
+@Repository
+public interface PostRepository extends JpaRepository<Post, Long> {
+
+    @Query("SELECT p FROM Post p WHERE p.author = :author_id")
+    List<Post> findByAuthor(@Param("author_id") Long authorId);
```
Comment (markonikolic-nqode) [inline src/main/java/com/internship/socialnetwork/repository/PostRepository.java]: @Param("author_id") not needed, just use the same value authorId in the query too

File: src/main/java/com/internship/socialnetwork/service/PostService.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+import com.internship.socialnetwork.dto.PostDto;
+import com.internship.socialnetwork.model.Post;
+
+import java.util.List;
+
+public interface PostService {
+
+    List<PostDto> getAll();
+
+    List<PostDto> getAllByAuthorId(Long id);
```
Comment (markonikolic-nqode) [inline src/main/java/com/internship/socialnetwork/service/PostService.java]: Pls rename parameter to authorId

File: src/main/java/com/internship/socialnetwork/service/impl/PostServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+@Service
+@RequiredArgsConstructor
+public class PostServiceImpl implements PostService {
+
+    private final PostRepository postRepository;
+
+    @Override
+    public Post getModel(Long id) {
+        return postRepository.findById(id).orElseThrow(() -> new NotFoundException("Post with id: " + id + " not found!"));
```
Comment (markonikolic-nqode) [inline src/main/java/com/internship/socialnetwork/service/impl/PostServiceImpl.java]: This string can be a constant, you can use String.format for interpolating id value

File: src/main/java/com/internship/socialnetwork/service/impl/PostServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    @Override
+    public List<PostDto> getAllByAuthorId(Long id) {
+        return postRepository.findByAuthor(id).stream().map(PostDto::convertToDto).toList();
+    }
+
+    @Override
+    public PostDto get(Long id) {
+        return PostDto.convertToDto(postRepository.findById(id).orElseThrow(() -> new NotFoundException("Post with id: " + id + " not found!")));
```
Comment (markonikolic-nqode) [inline src/main/java/com/internship/socialnetwork/service/impl/PostServiceImpl.java]: convertToDto can be a static import

File: src/main/java/com/internship/socialnetwork/service/impl/PostServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    @Override
+    public PostDto create(PostDto postDto) {
+        Post post = postDto.convertToModel();
```
Comment (markonikolic-nqode) [inline src/main/java/com/internship/socialnetwork/service/impl/PostServiceImpl.java]: post variable not needed

File: src/main/java/com/internship/socialnetwork/service/impl/PostServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+        Post post = getModel(id);
+        post.setText(postDto.getText());
+        post.setVideoPath(postDto.getVideoPath());
+        post.setImagePath(postDto.getImagePath());
+        return PostDto.convertToDto(postRepository.save(post));
+    }
+
+    @Override
+    public void delete(Long id) {
+        postRepository.findById(id).orElseThrow(() -> new NotFoundException("Post with id: " + id + " not found!"));
```
Comment (markonikolic-nqode) [inline src/main/java/com/internship/socialnetwork/service/impl/PostServiceImpl.java]: you don't need to search by id 2x, the method below also does find by id. You can put postRepository.delete(getModel(id)) instead

File: src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoMoreInteractions;
+import static org.mockito.Mockito.when;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+@ExtendWith(MockitoExtension.class)
+public class PostServiceImplTest {
```
Comment (markonikolic-nqode) [inline src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java]: public not needed

File: src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+
+    private final String text2 = "Text2";
+
+    @Mock
+    private PostRepository postRepository;
+
+    @InjectMocks
+    private PostServiceImpl postService;
+
+    private User createAuthor(Long id, String username) {
```
Comment (markonikolic-nqode) [inline src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java]: private methods should be below tests

File: src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+
+        // Verify that postRepository.findById(...) was called exactly once with userId2
+        verify(postRepository, times(1)).findById(id1);
+        verifyNoMoreInteractions(postRepository);
+    }
+
+    @Test
+    void shouldGetAllPosts() {
+        // Given: Mocking list of posts
+        List<Post> posts = new ArrayList<>();
```
Comment (markonikolic-nqode) [inline src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java]: List.of instead

File: src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+
+        // Verify that postRepository.findAll() was called exactly once
+        verify(postRepository, times(1)).findAll();
+        verifyNoMoreInteractions(postRepository);
+    }
+
+    @Test
+    void shouldGetAllPostsBySameAuthor() {
+        // Given: Mocking list of posts
+        List<Post> posts = new ArrayList<>();
```
Comment (markonikolic-nqode) [inline src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java]: The same here

File: src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+    @Test
+    void shouldCreatePost() {
+        // Given:
+        PostDto postDto = PostDto.builder()
+                .author(createAuthor(id1, text1))
+                .text(text1)
+                .imagePath(text2)
+                .build();
+
+        Post savedPost = Post.builder()
```
Comment (markonikolic-nqode) [inline src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java]: you can use createPost method here

File: src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+        when(postRepository.findById(id1)).thenReturn(Optional.of(post));
+
+        // When: Deleting the post
+        postService.delete(id1);
+
+        // Then: Verify that the post is deleted
+        assertEquals(0, postService.getAll().size());
+
+        // Verify that postRepository.deleteById(...) was called exactly once with the post ID
+        verify(postRepository, times(1)).deleteById(id1);
```
Comment (markonikolic-nqode) [inline src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java]: here you should verify findById too

File: src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java [hunk 15]
```diff
@@ -0,0 +1,203 @@
...
+        when(postRepository.findById(id1)).thenReturn(Optional.empty());
+
+        // When: Deleting the post
+        NotFoundException exception = assertThrows(NotFoundException.class, () -> postService.delete(id1));
+
+        // Then: Verify that NotFoundException is thrown with appropriate message
+        assertEquals("Post with id: 1 not found!", exception.getMessage());
+
+        // Verify that postRepository.deleteById(...) was not called
+        verify(postRepository, never()).deleteById(any());
```
Comment (markonikolic-nqode) [inline src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java]: the same here

File: src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java [hunk 16]
```diff
@@ -0,0 +1,203 @@
...
+    @Test
+    void shouldThrowNotFoundException_whenGetPost_ifPostDoesNotExist() {
+        // Given: Mocking the scenario where the post does not exist
+        when(postRepository.findById(id1)).thenReturn(Optional.empty());
+
+        // When: Getting post
+        NotFoundException exception = assertThrows(NotFoundException.class, () -> postService.get(id1));
+
+        // Then: Verify that NotFoundException is thrown with appropriate message
+        assertEquals("Post with id: 1 not found!", exception.getMessage());
```
Comment (markonikolic-nqode) [inline src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java]: Can be a constant

**Review by markonikolic-nqode — CHANGES_REQUESTED** [review]

File: src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java [hunk 17]
```diff
@@ -0,0 +1,203 @@
...
+    private PostRepository postRepository;
+
+    @InjectMocks
+    private PostServiceImpl postService;
+
+    private User createAuthor(Long id, String username) {
+        return new User() {{
+            setId(id);
+            setUsername(username);
+        }};
```
Comment (vkotarlic) [inline src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java]: Use one way to create objects in our case we asked for Builder and this should be used everywhere.

File: src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java [hunk 18]
```diff
@@ -0,0 +1,203 @@
...
+    }
+
+    private Post createPost(Long postId, Long authorId, String username){
+        return new Post() {{
+                setId(postId);
+                setAuthor(createAuthor(authorId, username));
+                setImagePath(text2);
+        }};
```
Comment (vkotarlic) [inline src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java]: Use one way to create objects in our case we asked for Builder and this should be used everywhere.

**Review by vkotarlic — CHANGES_REQUESTED** [review]

**Comment by markonikolic-nqode** [comment]
Pls fix conflicts and pending comments, also you have 2 commits with very similar message

File: src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java [hunk 19]
```diff
@@ -0,0 +1,207 @@
...
+
+    private final Long id1 = 1L;
+
+    private final Long id2 = 2L;
+
+    private final Long id3 = 3L;
+
+    private final String text1 = "Text1";
+
+    private final String text2 = "Text2";
```
Comment (vkotarlic) [inline src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java]: Do we really need to create variables named like this for numbers and random text

File: src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java [hunk 20]
```diff
@@ -0,0 +1,207 @@
...
+    private final String id1NotFound = "Post with id: 1 not found!";
+
+    @Mock
+    private PostRepository postRepository;
+
+    @InjectMocks
+    private PostServiceImpl postService;
+
+    @Test
+    void shouldGetPost(){
```
Comment (vkotarlic) [inline src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java]: Test name is not consistent with rest of method names in this test class. According to naming standard used below this should be named: `shouldReturnPost_whenGetPost_ifPostExists`

File: src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java [hunk 21]
```diff
@@ -0,0 +1,207 @@
...
+        // Then: Verify that NotFoundException is thrown with appropriate message
+        assertEquals(id1NotFound, exception.getMessage());
+
+        // Verify that postRepository.findById(...) was called exactly once with userId2
+        verify(postRepository, times(1)).findById(id1);
+        verifyNoMoreInteractions(postRepository);
+    }
+
+    @Test
+    void shouldGetAllPosts() {
```
Comment (vkotarlic) [inline src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java]: Same as above for `shouldGetPost`

File: src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java [hunk 22]
```diff
@@ -0,0 +1,207 @@
...
+        // Then: Verify that correct number of posts is returned
+        assertEquals(2, postDtos.size());
+
+        // Verify that postRepository.findAll() was called exactly once
+        verify(postRepository, times(1)).findAll();
+        verifyNoMoreInteractions(postRepository);
+    }
+
+    @Test
+    void shouldGetAllPostsBySameAuthor() {
```
Comment (vkotarlic) [inline src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java]: Same as above for `shouldGetPost`

File: src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java [hunk 23]
```diff
@@ -0,0 +1,207 @@
...
+        // Then: Verify that correct number of posts is returned
+        assertEquals(2, postDtos.size());
+
+        // Verify that postRepository.findByAuthor() was called exactly once
+        verify(postRepository, times(1)).findByAuthor(id1);
+        verifyNoMoreInteractions(postRepository);
+    }
+
+    @Test
+    void shouldCreatePost() {
```
Comment (vkotarlic) [inline src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java]: Same as above for `shouldGetPost`

File: src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java [hunk 24]
```diff
@@ -0,0 +1,207 @@
...
+
+        // Verify that postRepository.save(...) was called exactly once with any(Post.class) argument
+        verify(postRepository, times(1)).save(any(Post.class));
+
+        // Verify that there are no more interactions with postRepository
+        verifyNoMoreInteractions(postRepository);
+    }
+
+    @Test
+    void shouldDeletePost() {
```
Comment (vkotarlic) [inline src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java]: Same as above for `shouldGetPost`

File: src/main/java/com/internship/socialnetwork/service/impl/PostServiceImpl.java [hunk 25]
```diff
@@ -0,0 +1,62 @@
...
+
+import static com.internship.socialnetwork.dto.PostDto.convertToDto;
+
+@Service
+@RequiredArgsConstructor
+public class PostServiceImpl implements PostService {
+
+    private final PostRepository postRepository;
+
+    private final String notFoundExceptionMessage = "Post with id: %1$s not found!";
```
Comment (vkotarlic) [inline src/main/java/com/internship/socialnetwork/service/impl/PostServiceImpl.java]: If final (constant) this should be in upper case

**Review by vkotarlic — CHANGES_REQUESTED** [review]

**Review by markonikolic-nqode — CHANGES_REQUESTED** [review]
Pls fix conflicts before requesting a review. Also you have 2 commits with the identical message

File: src/main/java/com/internship/socialnetwork/controller/PostController.java [hunk 26]
```diff
@@ -0,0 +1,54 @@
...
+    @GetMapping("/{id}")
+    public ResponseEntity<PostDto> getOne(@RequestParam Long id) {
+        return ResponseEntity.ok(postService.get(id));
+    }
+
+    //ToDo: Create getAllPostsById in the user controller
+
+    @PostMapping
+    public ResponseEntity<PostDto> create(@Valid @RequestBody PostDto postDto) {
+        return ResponseEntity.status(201).body(postService.create(postDto));
```
Comment (markonikolic-nqode) [inline src/main/java/com/internship/socialnetwork/controller/PostController.java]: You don't have to hardcode 201, use this instead: ResponseEntity.status(HttpStatus.CREATED)

File: src/main/java/com/internship/socialnetwork/service/impl/PostServiceImpl.java [hunk 27]
```diff
@@ -0,0 +1,62 @@
...
+
+import static com.internship.socialnetwork.dto.PostDto.convertToDto;
+
+@Service
+@RequiredArgsConstructor
+public class PostServiceImpl implements PostService {
+
+    private final PostRepository postRepository;
+
+    private static final String NOT_FOUND_EXCEPTION_MESSAGE = "Post with id: %1$s not found!";
```
Comment (markonikolic-nqode) [inline src/main/java/com/internship/socialnetwork/service/impl/PostServiceImpl.java]: Constants should be above injected repos or services, just below class declaration

File: src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java [hunk 28]
```diff
@@ -0,0 +1,190 @@
...
+            .build();
+
+    private final Post otherPost = Post.builder()
+            .id(2L)
+            .text("other text")
+            .author(author)
+            .imagePath("otherPath")
+            .build();
+
+    private static final String ID_NOT_FOUND = "Post with id: %1$s not found!";
```
Comment (markonikolic-nqode) [inline src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java]: Pls move this below class declaration and change value to "Post with id: 1 not found!"

File: src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java [hunk 29]
```diff
@@ -0,0 +1,190 @@
...
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoMoreInteractions;
+import static org.mockito.Mockito.when;
+
+import java.util.List;
+import java.util.Optional;
+
+@ExtendWith(MockitoExtension.class)
+public class PostServiceImplTest {
+    private final User author = User.builder()
```
Comment (markonikolic-nqode) [inline src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java]: Pls use private methods to build entities here

File: src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java [hunk 30]
```diff
@@ -0,0 +1,190 @@
...
+
+        // Verify that postRepository.findById(...) was called exactly once with userId2
+        verify(postRepository, times(1)).findById(post.getId());
+        verifyNoMoreInteractions(postRepository);
+    }
+
+    @Test
+    void shouldReturnAllPosts_whenGetAll() {
+        // Given: Mocking list of posts
+        List<Post> posts = List.of(
```
Comment (markonikolic-nqode) [inline src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java]: posts variable not needed

File: src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java [hunk 31]
```diff
@@ -0,0 +1,190 @@
...
+
+        // Verify that postRepository.findAll() was called exactly once
+        verify(postRepository, times(1)).findAll();
+        verifyNoMoreInteractions(postRepository);
+    }
+
+    @Test
+    void shouldReturnAllPostsBySameAuthor_whenGetAllByAuthorId() {
+        // Given: Mocking list of posts
+        List<Post> posts = List.of(
```
Comment (markonikolic-nqode) [inline src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java]: posts variable not needed

**Review by markonikolic-nqode — CHANGES_REQUESTED** [review]

File: src/main/java/com/internship/socialnetwork/dto/PostDto.java [hunk 32]
```diff
@@ -0,0 +1,44 @@
...
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+import jakarta.validation.constraints.NotNull;
+
+@Data
+@AllArgsConstructor
+@Builder
+@NoArgsConstructor
```
Comment (vkotarlic) [inline src/main/java/com/internship/socialnetwork/dto/PostDto.java]: Please try to sort annotations in some order for example `@Data, @Builder, @NoArgs, @AllArgs`

File: src/main/java/com/internship/socialnetwork/model/Post.java [hunk 33]
```diff
@@ -13,13 +18,16 @@
 
 @Data
 @NoArgsConstructor
+@AllArgsConstructor
+@Builder
 @Entity
```
Comment (vkotarlic) [inline src/main/java/com/internship/socialnetwork/model/Post.java]: Same as above, but `@Entity` should always be on class

File: src/main/java/com/internship/socialnetwork/model/User.java [hunk 34]
```diff
@@ -20,6 +20,8 @@
 @Builder
 @AllArgsConstructor
 @NoArgsConstructor
+@AllArgsConstructor
+@Builder
 @Entity
 @Table(name = "`user`")
```
Comment (vkotarlic) [inline src/main/java/com/internship/socialnetwork/model/User.java]: Why we have `@Builder` two times

File: src/main/java/com/internship/socialnetwork/service/impl/PostServiceImpl.java [hunk 35]
```diff
@@ -0,0 +1,62 @@
...
+
+    @Override
+    public Post getModel(Long id) {
+        return postRepository.findById(id).orElseThrow(() -> new NotFoundException(String.format(NOT_FOUND_EXCEPTION_MESSAGE, id)));
```
Comment (vkotarlic) [inline src/main/java/com/internship/socialnetwork/service/impl/PostServiceImpl.java]: Please break line at `.orElseThrow()`, always try to break code so it is not crossing grey line in IntellJ

File: src/main/java/com/internship/socialnetwork/service/impl/PostServiceImpl.java [hunk 36]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    @Override
+    public List<PostDto> getAll() {
+        return postRepository.findAll().stream().map(PostDto::convertToDto).toList();
```
Comment (vkotarlic) [inline src/main/java/com/internship/socialnetwork/service/impl/PostServiceImpl.java]: Same as above

File: src/main/java/com/internship/socialnetwork/service/impl/PostServiceImpl.java [hunk 37]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    @Override
+    public List<PostDto> getAllByAuthorId(Long authorId) {
+        return postRepository.findByAuthor(authorId).stream().map(PostDto::convertToDto).toList();
```
Comment (vkotarlic) [inline src/main/java/com/internship/socialnetwork/service/impl/PostServiceImpl.java]: Same as above

File: src/main/java/com/internship/socialnetwork/service/impl/PostServiceImpl.java [hunk 38]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    @Override
+    public PostDto get(Long id) {
+        return convertToDto(postRepository.findById(id).orElseThrow(() -> new NotFoundException(String.format(NOT_FOUND_EXCEPTION_MESSAGE, id))));
```
Comment (vkotarlic) [inline src/main/java/com/internship/socialnetwork/service/impl/PostServiceImpl.java]: Same as above

File: src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java [hunk 39]
```diff
@@ -0,0 +1,190 @@
...
+        // Then: Verify that NotFoundException is thrown with appropriate message
+        assertEquals(String.format(ID_NOT_FOUND, post.getId()), exception.getMessage());
+
+        // Verify that postRepository.findById(...) was called exactly once with userId2
+        verify(postRepository, times(1)).findById(post.getId());
+        verifyNoMoreInteractions(postRepository);
+    }
+
+    @Test
+    void shouldReturnAllPosts_whenGetAll() {
```
Comment (vkotarlic) [inline src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java]: Test name is not consistent with rest of class, you are missing `_IfPostsExist`

File: src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java [hunk 40]
```diff
@@ -0,0 +1,190 @@
...
+        // Then: Verify that correct number of posts is returned
+        assertEquals(2, postDtos.size());
+
+        // Verify that postRepository.findAll() was called exactly once
+        verify(postRepository, times(1)).findAll();
+        verifyNoMoreInteractions(postRepository);
+    }
+
+    @Test
+    void shouldReturnAllPostsBySameAuthor_whenGetAllByAuthorId() {
```
Comment (vkotarlic) [inline src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java]: Same as above

File: src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java [hunk 41]
```diff
@@ -0,0 +1,190 @@
...
+        // Then: Verify that correct number of posts is returned
+        assertEquals(2, postDtos.size());
+
+        // Verify that postRepository.findByAuthor() was called exactly once
+        verify(postRepository, times(1)).findByAuthor(post.getAuthor().getId());
+        verifyNoMoreInteractions(postRepository);
+    }
+
+    @Test
+    void shouldCreatePost_whenCreate() {
```
Comment (vkotarlic) [inline src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java]: Same as above

File: src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java [hunk 42]
```diff
@@ -0,0 +1,190 @@
...
+
+        // Verify that postRepository.save(...) was called exactly once with any(Post.class) argument
+        verify(postRepository, times(1)).save(any(Post.class));
+
+        // Verify that there are no more interactions with postRepository
+        verifyNoMoreInteractions(postRepository);
+    }
+
+    @Test
+    void shouldDeletePost_whenDelete() {
```
Comment (vkotarlic) [inline src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java]: Same as above

**Review by vkotarlic — CHANGES_REQUESTED** [review]
Overall good job, keep it up! 😄

File: src/main/java/com/internship/socialnetwork/service/impl/PostServiceImpl.java (hunk 27 above)
Comment (markonikolic-nqode) [reply src/main/java/com/internship/socialnetwork/service/impl/PostServiceImpl.java]: ALso the message should bi renamed to POST_NOT_FOUND or POST_NOT_FOUND_(MSG, ERROR). Whatever you choose but then just follow the same format, not somewhere with _MSG and somewhere without it

File: src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java (hunk 28 above)
Comment (markonikolic-nqode) [reply src/test/java/com/internship/socialnetwork/service/PostServiceImplTest.java]: Now that you hardcoded 1 you don't need to call String.format when using this constant

File: src/test/java/com/internship/socialnetwork/service/impl/PostServiceImplTest.java [hunk 43]
```diff
@@ -0,0 +1,189 @@
...
+    @Test
+    void shouldThrowNotFoundException_whenGetPost_ifPostDoesNotExist() {
+        // Given: Mocking the scenario where the post does not exist
+        when(postRepository.findById(post.getId())).thenReturn(Optional.empty());
+
+        // When: Getting post
+        NotFoundException exception = assertThrows(NotFoundException.class, () -> postService.get(post.getId()));
+
+        // Then: Verify that NotFoundException is thrown with appropriate message
+        assertEquals(String.format(ID_NOT_FOUND), exception.getMessage());
```
Comment (markonikolic-nqode) [inline src/test/java/com/internship/socialnetwork/service/impl/PostServiceImplTest.java]: String.format not needed now that you have hardcoded userId

File: src/test/java/com/internship/socialnetwork/service/impl/PostServiceImplTest.java [hunk 44]
```diff
@@ -0,0 +1,189 @@
...
+    private final Post otherPost = createPost(2L);
+
+    @Mock
+    private PostRepository postRepository;
+
+    @InjectMocks
+    private PostServiceImpl postService;
+
+    @Test
+    void shouldReturnPost_whenGetPost_ifPostExists(){
```
Comment (markonikolic-nqode) [inline src/test/java/com/internship/socialnetwork/service/impl/PostServiceImplTest.java]: space needed: () {

**Review by markonikolic-nqode — CHANGES_REQUESTED** [review]

File: src/main/java/com/internship/socialnetwork/service/impl/PostServiceImpl.java [hunk 45]
```diff
@@ -0,0 +1,70 @@
...
+
+import java.util.List;
+
+import static com.internship.socialnetwork.dto.PostDto.convertToDto;
+
+@Service
+@RequiredArgsConstructor
+public class PostServiceImpl implements PostService {
+
+    private static final String NOT_FOUND_EXCEPTION_MESSAGE = "Post with id: %1$s not found!";
```
Comment (markonikolic-nqode) [inline src/main/java/com/internship/socialnetwork/service/impl/PostServiceImpl.java]: Should be Post with id %s not found! Also check other comment about rename

**Review by markonikolic-nqode — APPROVED** [review]

File: src/main/java/com/internship/socialnetwork/repository/PostRepository.java [hunk 46]
```diff
@@ -0,0 +1,16 @@
...
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.stereotype.Repository;
+
+import java.util.List;
+
+@Repository
+public interface PostRepository extends JpaRepository<Post, Long> {
+
+    @Query("SELECT p FROM Post p WHERE p.author = :authorId")
```
Comment (vkotarlic) [inline src/main/java/com/internship/socialnetwork/repository/PostRepository.java]: Did you check this? I think you need `p.author.id` here

File: src/test/java/com/internship/socialnetwork/service/impl/PostServiceImplTest.java [hunk 47]
```diff
@@ -0,0 +1,190 @@
...
+@ExtendWith(MockitoExtension.class)
+public class PostServiceImplTest {
+
+    private static final String NOT_FOUND_EXCEPTION_MSG = "Post with id 1 not found!";
+
+    private final User author = createUser();
+
+    private final Post post = createPost(1L);
+
+    private final Post otherPost = createPost(2L);
```
Comment (vkotarlic) [inline src/test/java/com/internship/socialnetwork/service/impl/PostServiceImplTest.java:36]: Do you need variables here, you can just call these methods where you need them

**Review by vkotarlic — CHANGES_REQUESTED** [review]

**Review by vkotarlic — APPROVED** [review]

**Pull request merged**